2024 05 09 raft cleanup

andrewbaptist opened the pull request:
This set of commits are minor refactoring of replica_raft.go. The bulk of the changes are to remove all the uses of non-deferred `Unlock`. There should not be any functional changes in these commits.

Release Note: None

Part of: https://github.com/cockroachdb/cockroach/issues/105366

Commit:
kvserver: use defer unlock in sendRaftMessage

Extract a method to enable the use of deferred lock release.

Release note: None

Part of: #105366

Commit:
kvserver: switch droppedMessages to atomic

Removes an unnecessary locking of the Replica.mu to updates a metrics
counter.

Release note: None

Commit:
kvserver: move unlock to defer in addUnreachableRemoteReplica

Release note: None

Part of: #105366

Commit:
kvserver: move updateProtectedState to own method

Allows the use of deferred unlock.

Release note: None

Part of: #105366

Commit:
kvserver: move maybeRefreshProposals to own method

Allows the use of deferred unlock.

Release note: None

Part of: #105366

Commit:
kvserver: create method createRaftState

Defer the release of the lock.

Part of: 105366

Release note: None

Commit:
kvserver: use withRaftLocked method

Allows the use of deferred unlocks.

Release note: None

Part of: #105366

Commit:
kvserver: use withRaftGroup instead of withRaftGroupLocked

Release note: None

Part of: #105366

Commit:
kvserver: extract clearing of the unreachable nodes

Extract into a method to allow the use of defer unlock.

Part of: #105366

Release note: None

Commit:
kvserver: defer unlock in maybeCoalesceHeartbeat

Release note: None

Part of: #105366

Commit:
kvserver: defer unlock in Replica.localMsgs

Part of: #105366

Release note: None

Commit:
kvserver: move the test check into the locked function

Part of: #105366

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/123886)
<!-- Reviewable:end -->

Commit:
kvserver: remove exclusion from replica_raft

As part of #107577 an exclusion was added to replica_raft.go. After this
PR this exclusion is no longer necessary.

Part of: #105366

Release note: None

Commit:
kvserver: extract function assertValidRaftCommand

This change extracts the assertion validation for `MsgApp` requests to
make it more clear that the assertions are only called on race paths.

Release note: None

Commit:
kvserver: call isInitialized instead of checking endKey

Release note: None

Commit:
kvserver: simplify maybeAcquireSnapshotMergeLock unlock behavior

This commit doesn't change the behavior of the method, but makes it more
explicit that the unlock calls will be specifically tied to the replicas
that are subsumed. This method intentionally "leaks" an unlocked raftMu,
therefore it is still necessary to add `nolint:deferunlockcheck`.

Part of: #105366

Release note: None
